### DISCUSSION TITLE
Supporting more python versions

### DISCUSSION THREAD

**Discussion by orbingol** [microsoft/vcpkg#17352] [Ideas]
I've recently discovered the "versions" feature of vcpkg and I think it is a great addition. The instructions on how to use this feature are posted [here](https://devblogs.microsoft.com/cppblog/take-control-of-your-vcpkg-dependencies-with-versioning-support/)

Due to some limitations in the dependent libraries, I usually need to use python one major version older than the latest release, e.g. the latest version is 3.9.x as of today but I need to use python 3.8.x. I've checked the vcpkg versions file (https://github.com/microsoft/vcpkg/blob/master/versions/p-/python3.json) and I see that vcpkg supports up to python v3.8.3 for v3.8.x series. However, it seems like python released v3.8.9 a few weeks ago (https://www.python.org/downloads/).

Wouldn't be great for vcpkg to support v3.8 and v3.9 versions together? What do you think about it?

**Comment by ras0219-msft** [comment]
Thanks for the suggestion! This is intended to be solved via registries other than Microsoft/vcpkg (see https://github.com/microsoft/vcpkg/discussions/17161 and https://github.com/microsoft/vcpkg/blob/master/docs/specifications/registries-2.md).

Because multiple minor versions of python cannot be built and installed side-by-side, it would require doubling our entire CI set to accommodate it. As you can imagine, this quickly gets out of hand as soon as you add 4 or 5 packages that require this treatment (32x the CI time). Therefore, we cannot properly support these variants in the public, curated catalog and instead encourage interested users to set up their own registries for these alternative versions.

**Reply by Hoikas** [reply]
A separate Python registry would be interesting. and I wouldn't mind investigating such a thing myself. The main caveat to this being that I don't think any ports in the official registry can build Python bindings against any python versions maintained in the python registry. Not sure about that, but if there's interest in a Python registry, I might be willing to try :smile:

**Reply by orbingol** [reply]
I understand the concerns @ras0219-msft. I was only thinking that it would be great to support the latest patch in each minor release. I don't think it is necessary to support, let's say, v3.8.3 after the release of v3.8.4 as the latter would be a bug-fix release with the same functionalities. Therefore, vcpkg would only support the latest patch of the 3.x minor releases. Of course, there is no need to support older minor releases, like Python v3.6 or deprecated ones, like v3.4.

@Hoikas, I have a feeling that a registry with all the minor and major releases would be very similar to the purpose of conda environments. It would be nice of course, but as explained above, I don't think it is necessary to support buggy older versions of Python. I might be missing a lot of things here by means of the registry concept :) but this is just my two cents.
